test/test_type_promotion.py

```diff
@@ -345,17 +345,42 @@ def test_create_bool_tensors(self, device):
 
     @float_double_default_dtype
     def test_result_type(self, device):
-        self.assertEqual(torch.result_type(torch.tensor(1, dtype=torch.int, device=device), 1), torch.int)
-        self.assertEqual(torch.result_type(1, torch.tensor(1, dtype=torch.int, device=device)), torch.int)
+        # Tensor against tensor
+        for dtypea in torch.testing.get_all_dtypes():
```

Comment:
You can simplify this test by using the `@dtypes` decorator like this:

```
@dtypes(*product(torch.testing.get_all_dtypes(), torch.testing.get_all_dtypes())
def test_result_type(self, device, dtypes):
  t = torch.tensor((0, 1), device=device, dtype=dtypes[1])
  t_scalar = torch.tensor(2, device=device, dtype=dtypes[1])
  scalar = b_scalar_tensor.item()
  args = (t, t_scalar, scalar)
  ...
```

Then I think you can simplify the body like this:

```
for a, b in product(args, args):
  try:
    result = a + b
  except RuntimeError:
    with self.assertRaises(RuntimeError):
      torch.promote_types(a, b)
    with self.assertRaises(RuntimeError):
      torch.result_type(a, b)
    continue

  self.assertEqual(result.dtype, torch.promote_types(a, b))
  self.assertEqual(result.dtype, torch.result_type(a, b)) 
```

I think something like that will ensure that `torch.promote_types`, `torch.result_type` and addition are consistent for scalars, tensors with zero dimensions, and tensors with non-zero dimensions.

What are your thoughts?

Reply:
1. I think the scalar part has to be singled out (as what has been done in the current code), because specifying single-element tensor dtype doesn't always take effects (e.g., uint8 would be seen as int64). See https://github.com/pytorch/pytorch/issues/42473#issuecomment-668792855 . Hence I think using `@dtypes` actually complicates the test code. What do you think?

2. I agree that having `+` in the test is a good idea.

Reply:
The scalar x scalar case is uninteresting, but doesn't result_type work if you have tensor x scalar? You could skip the promote_types check in that case (since it needs a.dtype, unlike the sketch I wrote above).

Reply:
What I meant is that the simplification of the code would actually make it more complicated in this situation, because all single-element tensors are seen as scalars, and scalars only have a handful dtypes available. The code would have a lot of branches that singles these cases out, which would be more difficult to read than the current flow.

Reply:
I think it would be more complicated if you're verifying the dtype is correct like you are now, but it's OK just to verify the dtype is consistent between add, promote_types, and result_types.

Reply:
I updated the code according to your code framework. I think the code does look better, but is not as simple as you might have thought. As the comment that I referred to above, the behavior is different for scalars (single-element tensors are treated as scalars), and has to be singled out when testing and thus adds complications because we are now doing all tests in a single loop. For example, the following two expressions give different results and this is intended:

```
In [11]: torch.tensor([190, 220], dtype=torch.int16) + torch.tensor(190, dtype=torch.int32)
Out[11]: tensor([380, 410], dtype=torch.int16)

In [12]: torch.tensor([190, 220], dtype=torch.int16) + torch.tensor([190, 220], dtype=torch.int32)
Out[12]: tensor([380, 440], dtype=torch.int32)
```

Reply:
I was thinking of a test like this:

```
    @dtypes(*itertools.product(torch.testing.get_all_dtypes(), torch.testing.get_all_dtypes()))
    def test_result_type(self, device, dtypes):
        t = torch.tensor((0, 1), device=device, dtype=dtypes[1])
        t_scalar = torch.tensor(2, device=device, dtype=dtypes[1])
        scalar = t_scalar.item()
        args = (t, t_scalar, scalar)

        for a, b in itertools.product(args, args):
            try:
                result = a + b
            except RuntimeError:
                with self.assertRaises(RuntimeError):
                    torch.promote_types(a, b)
                with self.assertRaises(RuntimeError):
                    torch.result_type(a, b)
                continue

        # Skips analyzing the scalar x scalar case
        if not isinstance(result, torch.Tensor):
            return

        # NOTE: promote_types assumes both arguments are tensors with 1+ dims
        if isinstance(a, torch.Tensor) and a.dim() != 0 and \
           isinstance(b, torch.Tensor) and b.dim() != 0:
            self.assertEqual(result.dtype, torch.promote_types(a.dtype, b.dtype))
        self.assertEqual(result.dtype, torch.result_type(a, b))
```

which avoids the additional complexity by skipping the separate promote types impl. What would you think of that?

Reply:
But wouldn't we miss the test on the resulting dtype of a tensor against a single-element tensor or a scalar?

Reply:
Yes and no. I think there's a trade-off we can discuss here.

This version doesn't test that `torch.promote_types` is consistent with the result dtype when a scalar or zero-dim tensor is involved in the computation, true, but we also don't expect `torch.promote_types` to always return a result consistent with the result type in those cases. This test does, however, validate that `torch.result_type` is always consistent with the addition. 

The test also doesn't attempt to validate using any "ground truth". It simply tests for consistency. That is, adding two torch.float32 tensors might return a long tensor, and the test wouldn't complain if `torch.promote_types` and `torch.result_type` agreed.

Validating that these functions are correct and not just consistent requires reimplementing, in part or whole, PyTorch's type promotion system in Python. This is tricky, as you point out, but I'm also not sure if it's worth doing. 

What are your thoughts?

Reply:
I think it's probably better to have the current additional tests preserved. This test is on `result_type`, not `+`, and I think it makes sense to ensure that with scalars, it should also produce expected results. What do you think?

Reply:
Wait I'm confused, the above does validate the consistency of `torch.result_type` and the dtype produced by addition when scalars and zero-dim tensors are involved. Can you elaborate on what you're thinking?

Reply:
Because this would go backward: The `+` operator is supposed to be consistent with `result_type` (not the other way around) and applies `result_type` (see https://github.com/pytorch/pytorch/blob/64a7939ee56fdc0a34b3de66aa9e6f3c96ef6860/aten/src/ATen/native/BinaryOps.cpp#L60 , https://github.com/pytorch/pytorch/blob/64a7939ee56fdc0a34b3de66aa9e6f3c96ef6860/aten/src/ATen/native/TensorIterator.cpp#L696-L697, and https://github.com/pytorch/pytorch/blob/64a7939ee56fdc0a34b3de66aa9e6f3c96ef6860/aten/src/ATen/native/TensorIterator.cpp#L89 ). Here we would like to verify that `result_type` gives the correct resulting dtype when it should: follow `promote_types` when both are tensors, and follow the special rules when one is tensor and the other is scalar.

Reply:
Right. But how does the test I presented fail to do this? I feel like I'm missing something here. It always checks `torch.result_type() == (a+b).dtype`, including in tensor + scalar cases, and it checks against `promote_types` when both inputs are 1D+ tensors, right?

Reply:
(We can also Slack about this if that would help.)

Reply:
The issue is that, for tensor + scalar, only `torch.result_type() == (a+b).dtype` is examined, but the correctness of `torch.result_type()` is not examined (this is a test case for `torch.result_type()`, not `+`). Ping me on Slack you still have doubts.

Reply:
As discussed offline, I have updated the test to keep the loop for tensor vs tensor and scalar vs scalar, and keep tensor vs scalar as spot check.